KSPCommunityFixes/BugFixes/UpgradesApplyToPrefabs.cs

```diff
@@ -0,0 +1,249 @@
+            FieldInfo partRefField = AccessTools.Field(typeof(PartListTooltip), nameof(PartListTooltip.partRef));
+
+            int i = 0;
+            for (; i < code.Count - 1; ++i)
+            {
+                if (code[i].opcode == OpCodes.Ldarg_1
+                    && code[i + 1].opcode == OpCodes.Ldfld && ReferenceEquals(code[i + 1].operand, prefabField))
+                {
+                    code[i + 1].operand = partRefField;
+                    code.RemoveAt(i);
+                    break;
+                }
+            }
+            ++i;
+            for (; i < code.Count - 1; ++i)
+            {
+                if (code[i].opcode == OpCodes.Ldarg_1
+                    && code[i + 1].opcode == OpCodes.Ldfld && ReferenceEquals(code[i + 1].operand, prefabField))
+                {
+                    code[i].opcode = OpCodes.Ldarg_0;
+                    code[i + 1].operand = partRefField;
+                    ++i;
+                }
+            }
+
+            return code;
+        }
+
+        static bool PartListTooltip_GetUpgradedPrimaryInfo_Prefix(ref AvailablePart aP, ref int maxLines, ref string __result)
+        {
+            // Replace with our substitute
+            Part part = GetSubstitutePart(aP);
+
+            List<string> list = new List<string>();
+
+            if (part.GetType().IsSubclassOf(typeof(Part)))
+            {
+                if (part is IModuleInfo moduleInfo)
+                    list.Add(moduleInfo.GetPrimaryField() + "\n");
+                else
+                    list.Add(part.drawStats().Trim());
+            }
+            for (int i = 0; i < part.Modules.Count; ++i)
+            {
+                if (part.Modules[i] is IModuleInfo moduleInfo)
+                    list.Add(moduleInfo.GetPrimaryField());
+            }
+            for (int i = 0; i < aP.resourceInfos.Count; ++i)
+            {
+                AvailablePart.ResourceInfo resourceInfo = aP.resourceInfos[i];
+                if (!string.IsNullOrEmpty(resourceInfo.primaryInfo))
+                    list.Add(resourceInfo.primaryInfo + "\n");
+            }
+
+            var stringBuilder = new System.Text.StringBuilder();
+            int max = Math.Min(maxLines, list.Count);
+            for (int i = 0; i < max; ++i)
+                stringBuilder.Append(list[i]);
+            __result = stringBuilder.ToString().Trim();
+
+            return false;
+        }
+    }
+
+    public class SubstitutePartRemoveFromDictOnDestroy : MonoBehaviour
+    {
+        private Dictionary<AvailablePart, Part> _dict;
+        private AvailablePart _key;
+
+        public void Setup(Dictionary<AvailablePart, Part> dict, AvailablePart key)
+        {
+            _dict = dict;
+            _key = key;
+        }
+
+        private void OnDestroy()
+        {
+            // Don't use Unity equality here
+            if (_dict != null && (object)_key != null)
```

Comment:
Use the `IsNotNullRef()` extension method here, those methods exists to avoid having to remember to comment these cases.

Reply:
ok!

Reply:
Others I implemented; this is no longer relevant because of the change to dict clearing on the root object.